src/kpro_batch.erl

```diff
@@ -0,0 +1,387 @@
+%%%   distributed under the License is distributed on an "AS IS" BASIS,
+%%%   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%%%   See the License for the specific language governing permissions and
+%%%   limitations under the License.
+%%%
+
+% Structures of kafka magic version 2.
+% For v0 and v1, see `kpro_batch_v01'.
+% In `kafka_protocol', there is no message version number in APIs,
+% the input pattern is used to figure out version number:
+% `[{Key, Value}]' : 0
+% `[{Timestamp, Key, Value}]': 1
+% `[#{headers => Headers, ts => Ts, key => Key, value => Value}]': 2
+
+-module(kpro_batch).
+
+-export([ decode/1
+        , encode/2
+        , encode_tx/4
+        , is_control/1
+        ]).
+
+-include("kpro_private.hrl").
+
+-type message() :: kpro:message().
+-type ts_type() :: kpro:timestamp_type().
+-type msg_ts() :: kpro:msg_ts().
+-type msg_input() :: kpro:msg_input().
+-type seqno() :: kpro:seqno().
+-type txn_ctx() :: kpro:txn_ctx().
+-type compress_option() :: kpro:compress_option().
+-type batch_input() :: kpro:batch_input().
+-type offset() :: kpro:offset().
+-type headers() :: kpro:headers().
+-type batch_meta() :: kpro:batch_meta().
+-define(NO_META, ?KPRO_NO_BATCH_META).
+
+%% @doc Encode a list of batch inputs into byte stream.
+%% Depending on the input pattern, if it is magic version 0 - 1
+%% call kpro_batch_v01 to do the work.
+%% If it is magic version 2, encode with dummy (non-transactional) batch meta.
+-spec encode(batch_input(), compress_option()) -> binary().
+encode([#{} | _] = Batch, Compression) ->
+  FirstSequence = -1,
+  NonTxn = #{ producer_id => -1
+            , producer_epoch => -1
+            },
+  iolist_to_binary(encode_tx(Batch, Compression, FirstSequence, NonTxn));
+encode(Batch, Compression) ->
+  iolist_to_binary(kpro_batch_v01:encode(Batch, Compression)).
+
+%% @doc Encode a batch of magic version 2.
+%% @end
+% RecordBatch =>
+%   FirstOffset => int64
+%   Length => int32
+%   PartitionLeaderEpoch => int32 # client set whatever
+%   Magic => int8 # The 17th byte as in v0 and v1
+%   CRC => int32
+%   Attributes => int16
+%   LastOffsetDelta => int32
+%   FirstTimestamp => int64
+%   MaxTimestamp => int64
+%   ProducerId => int64
+%   ProducerEpoch => int16
+%   FirstSequence => int32
+%   Records => [Record]
+-spec encode_tx(batch_input(), compress_option(), seqno(), txn_ctx()) ->
+        iodata().
+encode_tx([FirstMsg | _] = Batch, Compression, FirstSequence,
+          #{ producer_id := ProducerId
+           , producer_epoch := ProducerEpoch
+           }) ->
+  IsTxn = is_integer(ProducerId) andalso ProducerId >= 0,
+  FirstTimestamp =
+    case maps:get(ts, FirstMsg, false) of
+      false -> kpro_lib:now_ts();
+      Ts -> Ts
+    end,
+  EncodedBatch = encode_batch(Compression, FirstTimestamp, Batch),
+  EncodedAttributes = encode_attributes(Compression, IsTxn),
+  PartitionLeaderEpoch = -1, %% producer can set whatever
```

Comment:
Nit: comments trailing visible text (not just whitespace) should start with a single "%".